File: dsa/word-break/ts/word_break_v1.ts
function wordBreakR(
  str: string,
  idx: number,
  words: Array<string>,
): boolean {
  if (str.length === idx) return true;

  let prefix = "";

  for (let j = idx; j <= str.length; ++j) {
    prefix += str[j];

    if (words.includes(prefix) && wordBreakR(str, j + 1, words))
      return true;
  }

  return false;
}

export function wordBreak(str: string, words: Array<string>): boolean {
  return wordBreakR(str, 0, words);
}

File: dsa/word-break/py/wordbreakv1.py
def word_break_rec(s, i, words):
    if i == len(s):
        return 1

    n = len(s)
    prefix = ''

    for j in range(i, n):
        prefix += s[j]

        if prefix in words and word_break_rec(s, j + 1, words) == 1:
            return 1

    return 0


def word_break(s, words):
    return word_break_rec(s, 0, words)


if __name__ == '__main__':
    words = {'i', 'like', 'dsa'}
    print(word_break('dsadsa', words))
    # print(word_break('idsalike', words))
    # s = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'
    # words = {'a', 'aa', 'aaa', 'aaaa', 'aaaaa', 'aaaaaa', 'aaaaaaa', 'aaaaaaaa', 'aaaaaaaaa', 'aaaaaaaaaa'},
    # print(word_break(s, words))

# print(word_break('idsa', words))
#
# print(word_break('dsai', words))
#
# print(word_break('dsalike', words))
